File: src/main/java/com/example/sudoku/image/CellSplitter.java
package com.example.sudoku.image;

import java.util.ArrayList;
import java.util.List;

import org.opencv.core.Mat;
import org.opencv.core.Rect;
import org.opencv.core.Size;
import org.opencv.imgproc.Imgproc;

public class CellSplitter {

    public List<Mat> splitCells(Mat warped) {
        int size = warped.rows();
        int cellSize = size / 9;
        int margin = cellSize / 10; // crop margin inside cell
        List<Mat> cells = new ArrayList<>();

        for (int row = 0; row < 9; row++) {
            for (int col = 0; col < 9; col++) {
                int x = col * cellSize + margin;
                int y = row * cellSize + margin;
                int width = cellSize - 2 * margin;
                int height = cellSize - 2 * margin;
                Rect roi = new Rect(x, y, width, height);
                Mat cell = new Mat(warped, roi);
                Mat resized = new Mat();
                Imgproc.resize(cell, resized, new Size(28, 28));
                cells.add(resized);
            }
        }
        return cells;
    }
}

File: python/cell_splitter.py
"""Split a warped square board image into 81 cell images."""
def split_cells_from_array(warped):
    import cv2

    size = warped.shape[0]
    cell_size = size // 9
    margin = cell_size // 10
    cells = []
    for row in range(9):
        for col in range(9):
            x = col * cell_size + margin
            y = row * cell_size + margin
            w = cell_size - 2 * margin
            h = cell_size - 2 * margin
            roi = warped[y:y+h, x:x+w]
            resized = cv2.resize(roi, (28, 28))
            cells.append(resized)
    return cells
